clang/lib/CodeGen/CGCoroutine.cpp

```diff
@@ -129,7 +130,14 @@ static SmallString<32> buildSuspendPrefixStr(CGCoroData &Coro, AwaitKind Kind) {
   return Prefix;
 }
 
-static bool memberCallExpressionCanThrow(const Expr *E) {
+static bool ResumeExprCanThrow(const CoroutineSuspendExpr &S) {
+  const Expr *E = S.getResumeExpr();
+
+  // If the return type of await_resume is not void, get the CXXMemberCallExpr
+  // from its subexpr.
+  if (const auto *BindTempExpr = dyn_cast<CXXBindTemporaryExpr>(E)) {
+    E = BindTempExpr->getSubExpr();
+  }
```

Comment:
Such pattern match doesn't smell good. How about looking into its children recursively if we find `E` is not CXXMemberCallExpr?

Reply:
I find this a little overkill for what we need. The old code here didn't intend to handle more cases either because it just expected the AST to be in a specific shape (`CXXMemberCallExpr`). This fix is merely just to amend its functionality to work on a temporary expression as well should there be a return value to be discarded. It should be clear intent here that we don't care about anything else. (Ideally, I want an assertion here that it's either of the two `Expr` types)

Reply:
I think we are only targeting the resume part of `init_suspend` here, so the cases are limited. Can we change the function name to something more specific, like `InitResumeExprCanThrow`?

Reply:
I think we are only targeting the resume part of init_suspend here, so the cases are limited. Can we change the function name to something more specific, like `InitResumeExprCanThrow`?

Reply:
Maybe it is not a such big task as you imaged, you can take a look at `Sema::checkFinalSuspendNoThrow` and we should be able to make a simpler change than that.

The reason why I don't like the current pattern is that, every time I wrote:

```
if (auto *WantedExpressionType = isa<AExpreType>(E))
     E = WantedExpressionType->subExpr();
...
```

I found I always missed some situations.

Reply:
I guess you are correct that this is easy to miss some cases. In my case I missed the fact that the destructor of the `BindTempExpr->getTemporary()` can throw if marked `noexcept(false)`. I updated the PR to cover this case and including corresponding tests. 

Next up I will try to see how many cases I'll need to cover in a recursive check on `children()` like the one in `Sema::checkFinalSuspendNoThrow`.